### PR TITLE
More double backwards support for pooling, unpooling, padding

### PR DESCRIPTION
**Pull request by gchanan**
Implements double backwards for the following nn modules:
pooling: AdaptiveAvgPool1d, AdaptiveAvgPool2d, FractionalMaxPool2d
unpooling: MaxUnpool1d, MaxUnpool2d
padding: ReplicationPad2d, ReplicationPad3d, ReflectionPad2d.

### CODE REVIEW

**Comment by gchanan** [comment]
looks like a python2 issue.

File: torch/nn/_functions/thnn/auto.py [hunk 1]
```diff
@@ -333,4 +332,6 @@ def default_double_backwards_fn(ctx, *grad_params):
             _all_functions.append(backward_cls)
 
 
+from .auto_double_backwards import double_backwards_fns
```
Comment (apaszke) [inline torch/nn/_functions/thnn/auto.py]: It should be possible to keep this import at the top, right?

File: torch/nn/_functions/thnn/auto_double_backwards.py [hunk 2]
```diff
@@ -106,6 +106,27 @@ def logsoftmax_double_backwards(ctx, ggI):
     return gI, ggO, None, None, None, None
 
 
+def reflectionpad2d_double_backwards(ctx, ggI):
+    gI = None
+    ggO = torch.nn._functions.thnn.auto.ReflectionPad2d.apply(ggI, *ctx.additional_args)
```
Comment (apaszke) [inline torch/nn/_functions/thnn/auto_double_backwards.py]: It might be possible to shorten the chain (say start with `_functions` or `thnn`)

File: torch/nn/_functions/thnn/auto.py (hunk 1 above)
Comment (gchanan) [reply torch/nn/_functions/thnn/auto.py]: yes, will do.

File: torch/nn/_functions/thnn/auto_double_backwards.py (hunk 2 above)
Comment (gchanan) [reply torch/nn/_functions/thnn/auto_double_backwards.py]: played around with it, doesn't seem possible (at least in python 2).
